src/app/EventManagement.cpp

```diff
@@ -748,6 +750,59 @@ CHIP_ERROR EventManagement::FetchEventsSince(TLVWriter & aWriter, const ObjectLi
     return err;
 }
 
+CHIP_ERROR EventManagement::RemoveInvalidFabricCB(const TLV::TLVReader & aReader, size_t aDepth, void * apContext)
+{
+    FabricIndex * invalidFabricIndex = static_cast<FabricIndex *>(apContext);
+
+    TLVReader event;
+    TLVType tlvType;
+    TLVType tlvType1;
+    event.Init(aReader);
+    ReturnErrorOnFailure(event.EnterContainer(tlvType));
+    ReturnErrorOnFailure(event.Next());
+    ReturnErrorOnFailure(event.EnterContainer(tlvType1));
+    ReturnErrorOnFailure(event.Next());
+
+    uint8_t fabricIndex = 0;
+    while (CHIP_NO_ERROR == event.Next())
+    {
+        if (event.GetTag() == TLV::ProfileTag(kEventManagementProfile, kFabricIndexTag))
+        {
+            ReturnErrorOnFailure(event.Get(fabricIndex));
+            if (fabricIndex == *invalidFabricIndex)
+            {
+                uint8_t * dataPtr = const_cast<uint8_t *>(event.GetReadPoint());
+                dataPtr           = dataPtr - 1;
```

Comment:
The pointer arithmetic here should be checked against wraparound in the circular storage.  Something along the lines of 

```
    CHIPCircularTLVReader readBuffer = static_cast<CHIPCircularTLVReader *>(event.mBackingStore);
     if (readBuffer->GetQueue() != dataPtr) 
        dataPtr = dataPtr - 1;
    else
        dataPtr = readBuffer->GetQueue() +readBuffer->GetQueueSize() - 1;
```

This of course depends on being able to access the TLVBackingStore which we did not include.  So, perhaps there is a better way than my proposed hack that would save the ptr before starting this nightmare.  Perhaps saving a read point from line 763 would provide you a suitable place?

Reply:
if saving a read point from 763, it seems we have to manual calculate how much bytes from container place to fabric index, which could have event path, event number, priority, system timestamp or epoch timestamp, dynamic system/epoch timestamp if it has.....

Reply:
maybe as you proposed, we need to get accessor for  TLVBackingStore here.